### PR TITLE
feat(teams): endpoint to update membership roles

### PR DESCRIPTION
**Pull request by christyjacob4**
## What does this PR do?
Adds a new endpoint PATCH '/v1/teams/:teamId/memberships/:membershipId' to allow users to modify roles previously granted to team members.

Also created a new event to trigger in this case `teams.memberships.update`

## Test Plan
Added new tests

## Related PRs and Issues
NA

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
Yes

### CODE REVIEW

File: app/controllers/api/teams.php [hunk 1]
```diff
@@ -471,6 +471,83 @@
...
+
+        $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::$roles);
+        $isAppUser = Auth::isAppUser(Authorization::$roles);
+        $isOwner = false;
+
+        foreach ($memberships as $member) {
+            if ($member->getAttribute('userId') == $user->getId() && \in_array('owner', $member->getAttribute('roles', []))) {
+                $isOwner = true;
+            }
+        }
```
Comment (eldadfux) [inline app/controllers/api/teams.php:528]:
```suggestion
        $isOwner = Authorization::isRole('team:'.$team->getId().'/owner');;
```

This should work better, and helps us avoid fetching the memberships list.

File: app/controllers/api/teams.php [hunk 2]
```diff
@@ -471,6 +471,83 @@
...
+
+        $memberships = $projectDB->getCollection([
+            'limit' => 2000,
+            'offset' => 0,
+            'filters' => [
+                '$collection='.Database::SYSTEM_COLLECTION_MEMBERSHIPS,
+                'teamId='.$team->getId(),
+            ],
+        ]);
```
Comment (eldadfux) [inline app/controllers/api/teams.php:518]:
```suggestion
```

**Review by eldadfux — CHANGES_REQUESTED** [review]

**Review by TorstenDittmann — COMMENTED** [review]
We should make sure a team doesn't become ownerless.

Meaning, it shouldn't be possible to be the last `owner` and remove the owner role from yourself.

**Comment by eldadfux** [comment]
> We should make sure a team doesn't become ownerless.
> 
> Meaning, it shouldn't be possible to be the last `owner` and remove the owner role from yourself.

I think I would leave this to be handled on client and not force a specific flow. Even when we create teams from the console they start without any members. Using the servers SDKs you can manipulate the team membership or even use cloud functions to delete orphaned teams.

**Comment by TorstenDittmann** [comment]
> > We should make sure a team doesn't become ownerless.
> > Meaning, it shouldn't be possible to be the last `owner` and remove the owner role from yourself.
> 
> I think I would leave this to be handled on client and not force a specific flow. Even when we create teams from the console they start without any members. Using the servers SDKs you can manipulate the team membership or even use cloud functions to delete orphaned teams.

Fair point 👍🏻

File: app/controllers/api/teams.php [hunk 3]
```diff
@@ -471,6 +471,83 @@
         ;
     });
 
+App::patch('/v1/teams/:teamId/memberships/:membershipId')
+    ->desc('Update Membership Roles')
+    ->groups(['api', 'teams'])
+    ->label('event', 'teams.memberships.update')
+    ->label('scope', 'public')
+    ->label('sdk.auth', [APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_KEY])
```
Comment (lohanidamodar) [inline app/controllers/api/teams.php:479]: I believe `APP_AUTH_TYPE_JWT` is also required?

**Review by lohanidamodar — CHANGES_REQUESTED** [review]
Looks good, just a minor change.

File: app/controllers/api/teams.php (hunk 3 above)
Comment (eldadfux) [reply app/controllers/api/teams.php:479]: Good catch!

File: app/controllers/api/teams.php (hunk 1 above)
Comment (christyjacob4) [reply app/controllers/api/teams.php:528]: Oh I wasn't sure this was ready!

File: app/controllers/api/teams.php (hunk 3 above)
Comment (christyjacob4) [reply app/controllers/api/teams.php:479]: 👍

**Pull request merged**